lib/HLSL/HLOperationLower.cpp

```diff
@@ -4421,6 +4421,40 @@ static Value* SkipAddrSpaceCast(Value* Ptr) {
   return Ptr;
 }
 
+// For known non-groupshared, verify that the destination param is valid
+void ValidateAtomicDestination(CallInst *CI, HLObjectOperationLowerHelper *pObjHelper) {
+  Value *dest = CI->getArgOperand(HLOperandIndex::kInterlockedDestOpIndex);
+  GetElementPtrInst *gep = nullptr;
+
+  // If we encounter a gep, we may provide a more specific error message
+  while (isa<GetElementPtrInst>(dest)) {
+    gep = cast<GetElementPtrInst>(dest);
+    dest = gep->getPointerOperand();
+  }
+
+  // Confirm that dest is a UAV
+  if (CallInst *destCall = dyn_cast<CallInst>(dest)) {
+    hlsl::HLOpcodeGroup group = hlsl::GetHLOpcodeGroup(destCall->getCalledFunction());
+    while (group == HLOpcodeGroup::HLSubscript) {
+      dest = destCall->getArgOperand(HLOperandIndex::kSubscriptObjectOpIdx);
+      if (!isa<CallInst>(dest))
```

Comment:
I'm not sure what scenarios are possible here.  Could there be subscript->something else->subscript?  What's this operand when not another subscript?  What scenario has multiple layers of subscript operators?

I guess the general concern is about doing this on the HL ops.  Could it be performed before/during CodeGen (earlier is always better for the error reporting), or is that even more complicated?  Or could checking the DXIL translated ops simplify the check?

Reply:
Obviously, there is already some checking earlier, and a bit later.  So perhaps this is just about my uncertainty around what kinds of things are possible here that may need drilling through or not.

Reply:
The while loop over subscripts that I think is troubling you is needed to handle cases such as this:
https://github.com/microsoft/DirectXShaderCompiler/blob/50c0199c17b0b7530cbc0bdddadc3697eaece695/tools/clang/test/HLSLFileCheck/hlsl/intrinsics/atomic/atomic_structuredbuf_i64.hlsl#L36

Someday I'd like to refactor this code such that this is detectable in codegen, but it's not simple. The information to determine these errors is not readily available there and the errors were previously generated during this stage because of that. I decided not to take on that task for now, but we will have to move all this error checking to earlier eventually.

Reply:
I wonder about more complex cases, such as ones with other types of accesses in between.  Like, what does `simpArrBuf[a].ss[b].longEnding[c][d].x` end up looking like (assuming `d` is defined, and the matrix type is uint)?  Is it subscripts all the way down?  There's a matrix subscript here, but I guess that's an HLSubscript here too at this point.  What about the intermediate struct member, same?

I figured I might as well experiment to see what it looks like just before dxilgen, and I see:
`hl.subscript.[] -> GEP -> GEP -> GEP -> GEP -> hl.subscript.colMajor[] -> GEP -> hl atomic op`
Perhaps that's not what it looks like when it reaches this particular code though?  Does this code handle that gracefully as well?

Reply:
This would be simpler after more lowering in valid usages because the lowered atomic op would have a direct pointer to the resource. However, that lowering is performed dependent on the destination argument being a subscripted resource. If it isn't, lowering is missed entirely leaving an unlowered hl intrinsic. That leaves this as the best place to do it even though it requires some "drilling"